packages/openapi3/src/schema-emitter.ts

```diff
@@ -653,6 +660,180 @@ export class OpenAPI3SchemaEmitter extends TypeEmitter<
     }
   }
 
+  #attachXmlObjectForScalarOrModel(
+    program: Program,
+    prop: Scalar | Model,
+    emitObject: OpenAPI3Schema,
+  ) {
+    const xmlObject: OpenAPI3XmlSchema = {};
+
+    // Resolve XML name
+    const xmlName = resolveEncodedName(program, prop, "application/xml");
+    if (xmlName !== prop.name) {
+      xmlObject.name = xmlName;
+    }
+
+    // Get and set XML namespace if present
+    const currNs = getNs(program, prop);
+    if (currNs) {
+      xmlObject.prefix = currNs.prefix;
+      xmlObject.namespace = currNs.namespace;
+    }
+
+    // Attach xml schema to emitObject if not empty
+    if (Object.keys(xmlObject).length !== 0) {
+      emitObject.xml = xmlObject;
+    }
+  }
+
+  #attachXmlObjectForModelProperty(
+    program: Program,
+    prop: ModelProperty,
+    emitObject: OpenAPI3Schema,
+    ref?: Record<string, any>,
+  ) {
+    const xmlObject: OpenAPI3XmlSchema = {};
+
+    const isXmlModel = this.#isXmlModelChecker(program, prop.model!, []);
+    if (!isXmlModel) {
+      return;
+    }
+
+    // Resolve XML name
+    const xmlName = resolveEncodedName(program, prop, "application/xml");
+    const jsonName = resolveEncodedName(program, prop, "application/json");
+    if (xmlName !== prop.name && xmlName !== jsonName) {
+      xmlObject.name = xmlName;
+    }
+
+    // Get and set XML namespace if present
+    const currNs = getNs(program, prop);
+    if (currNs) {
+      xmlObject.prefix = currNs.prefix;
+      xmlObject.namespace = currNs.namespace;
+    }
+
+    // Set XML attribute if present
+    if (isAttribute(program, prop)) {
+      if (prop.type?.kind === "Model") {
+        reportDiagnostic(program, {
+          code: "invalid-property-type",
```

Comment:
those errors can be a bit more specific here, like say that xml attributes can only be primitive types in openapi3